File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.py
n = int(input())

def solution(n):
    # dp[i][j] i번자리수일 때 마지막 수가 j일 때 만들어지는 계단수의 개수
    inf = int(1e9)
    dp = [[-inf for _ in range(10)] for _ in range(n + 1)]
    mod = 1000000000
    dp[0][0] = 0
    for i in range(1, n + 1):
        if i == 1:
            for j in range(10):
                if j == 0:
                    dp[i][j] = 0
                else:
                    dp[i][j] = 1
        else:
            for j in range(10):
                if j == 0:
                    dp[i][j] = (dp[i - 1][j + 1]) % mod
                elif j == 9:
                    dp[i][j] = (dp[i - 1][j - 1]) % mod
                else:
                    dp[i][j] = (dp[i - 1][j - 1] + dp[i - 1][j + 1]) % mod
                    
    return (sum(dp[n])) % mod
        
print(solution(n))

File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.java

import java.io.*;
import java.util.*;

class Solution {
    public long solution(int n) {

        final long mod = 1000000000;

        long[][] dp = new long[n + 1][10];

        for (int j = 1; j < 10; j++) {
            dp[1][j] = 1;
        }

        for (int i = 2; i < n + 1; i++) {
            for (int j = 0; j < 10; j++) {
                if (j == 0) {
                    dp[i][j] = dp[i - 1][j + 1] % mod;
                } else if (j == 9) {
                    dp[i][j] = dp[i - 1][j - 1] % mod;
                } else {
                    dp[i][j] = (dp[i - 1][j - 1] + dp[i - 1][j + 1]) % mod;
                }
            }
        }

        long val = 0;
        for (int j = 0; j < 10; j++) {
//          System.out.printf("% d", dp[i][j]);
            val += dp[n][j] % mod;
        }
//                System.out.println();
        return val % mod;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        Solution sol = new Solution();
        System.out.println(sol.solution(n));

    }
}
